Update Precourse (#105)
Ada-Developers-Academy/ada-precourse-v2: intro-to-python/functions.md

Reviewer: That makes sense to me, I took a peek through and agree that though there is a bit of drift, having the videos in Functions is still helpful since the content is mostly in alignment.

def my_func_with_args(arg1, arg2):
    x = arg1
    y = arg2
    difference = arg1 - arg2
    print(difference)

# Calling the function with arguments
my_func_with_args(1, 2)
my_func_with_args(3, 2)

```

### Why Functions?
### Why Write and Use Functions?

In the last lesson we wrote a series of conditionals to validate and test user input for the "Guess the Number" game.  At this point, we may be looking at the code and thinking, this looks good, it works, why add functions?  
In the "Conditionals" lesson we wrote a series of conditionals to validate and test user input for the "Guess the Number" game.  At this point, we may be looking at the code and thinking, "this looks good, it works, why do we need functions"?  

Lets say we want to play a different game after we finish playing Guess the Number.  We could just add the code for the new game after the code that's in the file right now, but then what if we want to change the order of the games?  

> We're now looking at moving around big code blocks.  Then if we change our minds and want to move it back, or add another game (and so on) things quickly get messy. 

Author: This doesn't seem to illustrate the power of using functions to me. Even if you have functions, moving around large blocks of code can be cumbersome. I wrote out a different scenario that ties into keeping code DRY, which is introduced a few lines below.

> Imagine we'd like a user to be able to play the game more than once. Without functions, the code as it is currently written requires that we re-write the logic every time a user plays the game. If we want to allow the user to play the game 5 times, then we would need to write the same block of code 5 times. 

Author: Ah, yeah by "re-write" I meant copied again, like literally type the same code again. Let me think of a different way to word this.

> Imagine we'd like a user to be able to play the game more than once. Without functions, the code as it is currently written requires that we re-write the logic every time a user plays the game. If we want to allow the user to play the game 5 times, then we would need to write the same block of code 5 times. 

Author: How about "copy" instead of "re-write"?

Imagine we'd like a user to be able to play the game more than once. Without functions, the code as it is currently written requires that we **copy** the logic every time a user plays the game. If we want to allow the user to play the game 5 times, then we would need to write the same block of code 5 times.

> Imagine we'd like a user to be able to play the game more than once. Without functions, the code as it is currently written requires that we re-write the logic every time a user plays the game. If we want to allow the user to play the game 5 times, then we would need to write the same block of code 5 times. 

Reviewer: Copy is much clearer to me, I like that.

</details>

> With this example, we can see that calling a function allows us to run the same code multiple times without repeating a large block of code, which makes our code concise.  This code could be even more concise by using a loop, which we'll cover in a future lesson.  Run `game.py` to play the Guess the Number game.

Author: Ok good to know. Almost all of Precourse is written with 2 spaces so I stuck with it. 

I'm thinking that we can wrap up the content changes and then afterwards do one PR with only a change to make the double space a single space if that works. Don't want to bury any important content changes among a million space changes.

> With this example, we can see that calling a function allows us to run the same code multiple times without repeating a large block of code, which makes our code concise.  This code could be even more concise by using a loop, which we'll cover in a future lesson.  Run `game.py` to play the Guess the Number game.

Reviewer: Agreed 👍

## The DRY Principle

> "DRY stands for Don’t Repeat Yourself. It is a software development principle aiming to reduce code duplication which can lead to a poor refactoring and a poor maintenance. In other word, if you have to copy and paste the same exact lines of codes which provide the same functionality, then chances are that your codes don’t follow the DRY at all." [source](https://medium.com/@Ialimijoro/the-dry-principle-and-why-you-should-use-it-f02435ae9449) <br>

Author: This quote comes from a blog that isn't necessarily a robust resource and the quote is grammatically incorrect. I don't think we need to source an external resource so I just summarized DRY.

We will start our game by reading in a letter from the user, similar to how we read in numbers with `guess_the_number`.  We will call this function `get_letter_from_user`.

<!-- >>>>>>>>>>>>>>>>>>>>>> BEGIN CHALLENGE >>>>>>>>>>>>>>>>>>>>>> -->
<!-- Replace everything in square brackets [] and remove brackets  -->

### !challenge

* type: code-snippet
* language: python3.6
* id: c751d441-f566-4ab7-acc1-390b8710213e
* title: get_letter_from_user
* points: 1
* topics: python, python-functions

##### !question

> This function is very similar to `get_number_from_user`.  

Author: Deleted this line because we haven't discussed this function and it doesn't get introduced until the Lists lesson

1.  First, we will need to use `input` to get a string from the user and store it in a variable.  
1.  Second, we need to check to see if the input is valid.  

>     * In `get_number_from_user`, we used `isnumeric()` to ensure that the input was a number.  In this function, we need to check to see if the input is a letter, and if the input contains only one letter.

Author: The function where we use isnumeric is actually called guess_the_number. We haven't introduced get_number_from_user yet.

>       - `letter_from_user.isalpha()` will return `True` if the string variable `letter_from_user` contains only alphabetical characters

Reviewer: ```suggestion
      - If we have a variable which holds our input named `letter_from_user` then `letter_from_user.isalpha()` will return `True` if the string variable `letter_from_user` contains only alphabetical characters.
```

> We can add debugging print statements to our function so that we can certain values in our terminal if our logic isn't working as expected. For example, if we are curious about what the length of the user input is, we could add `print (len(letter))` in our function. After we have debugged an issue, we should be sure to remove the debugging print statement so that our code doesn't become cluttered with debugging code.

Reviewer: ```suggestion
We can add debugging print statements to our function so that we can see certain values in our terminal if our logic isn't working as expected. For example, if we are curious about what the length of the user input is, we could add `print (len(letter))` in our function. After we have debugged an issue, we should be sure to remove the debugging print statement so that our file doesn't become cluttered with debugging code.
```

> <summary>Once you've written `get_letter_from_user`, compare what you have to what we have in `snowman.py` so far.</summary>

Reviewer: ```suggestion
<summary>Once you've written <code>get_letter_from_user</code>, compare what you have to what we have in <code>snowman.py</code> so far.</summary>
```

```python

# Our hidden word 
SNOWMAN_WORD = "broccoli"

def get_letter_from_user():
    letter = input("Please enter a letter > ")

>     print("hey")

Reviewer: Is leaving in this print statement intentional?

Author: Nope, I copy and pasted code from a project, thanks for catching this!

    if len(letter) > 1 or not letter.isalpha():
        print("Invalid letter please enter a single character.")
    
    return letter

# Invoke the function the we wrote
get_letter_from_user()

```
</details>

> We are now going to work on the main `snowman` function (the function we will call when we want to play the game Snowman).  This function is  similar to `guess_the_number`.

Reviewer: The Snowman section is a little long, I'm wondering if it would help to visually break things up by adding a heading here that separates the `get_letter_from_user` work from the beginning of the `snowman` function work?

Author: Yeah that makes sense. We already have "Step 1: Guess a letter" so I can add "Step 2: Check for letter in SNOWMAN_WORD". What do you think?

Reviewer: Sounds good 👍

The only check we have to do on the letter is check if it's in the word or not.  To do that, we're going to use the python keyword `in`.  The syntax for `in` is `thing1 in thing2`.  This expression evaluates to `True` if `thing2` contains `thing1`, and `False` otherwise.  We can use it in a conditional expression as `if thing1 in thing2:`.

Our requirements for the `snowman` function are:

> 1.  First, use `get_letter_from_user` to get a letter

Reviewer: Since the following steps don't start with an ordering word, I feel like they match better if we remove "First" here.

```suggestion
1.  Use `get_letter_from_user` to get a letter
```

1.  Check if the letter is in `SNOWMAN_WORD` 
    - print "Letter found" if the letter is in `SNOWMAN_WORD`.
    - print "Letter not found" if it's not in `SNOWMAN_WORD`.

> 1.  If the letter is in `SNOWMAN_WORD` return `True` otherwise return `False` if it's not in `SNOWMAN_WORD`.

Reviewer: ```suggestion
1.  Return `True` if the letter is in `SNOWMAN_WORD`, otherwise return `False` if it's not in `SNOWMAN_WORD`.
```

Use the outlined requirements above to write the `snowman` function in `snowman.py`. We should call the `snowman` function and execute the file to check that our logic is working as we expect it to.

<br/>

<details>

> <summary>Once you've written `snowman`, compare what you have to what we have in `snowman.py` so far.</summary>

Reviewer: ```suggestion
<summary>Once you've written `snowman`, compare what you have to our sample code in `snowman.py` so far.</summary>
```

            # Act
            answer = p.snowman()

        # Assert
        assert re.match('Letter found', mock_stdout.getvalue(), flags=re.IGNORECASE)
        assert answer # True if it's in the snowman word

    @patch('sys.stdout', new_callable=io.StringIO)
    def test_p_four(self, mock_stdout):
        # Arrange
        input_letter = "e"
        with unittest.mock.patch('builtins.input', return_value=input_letter):
            # Act
            answer = p.snowman()

        # Assert
        assert re.match('Letter not found', mock_stdout.getvalue(), flags=re.IGNORECASE)
        assert answer == False
# Since the snowman function calls get_letter_from_user, we should test that snowman works as we expect it to by calling it
snowman()

```
</details>

##### !end-tests

<!-- other optional sections -->
<!-- !hint - !end-hint (markdown, hidden, students click to view) -->
<!-- !rubric - !end-rubric (markdown, instructors can see while scoring a checkpoint) -->
<!-- !explanation - !end-explanation (markdown, students can see after answering correctly) -->

### !end-challenge

<!-- ======================= END CHALLENGE ======================= -->
## Summary

> Let's go back to the questions proposed in the introduction to this lesson.  What do functions add to our code?  

Author: This isn't a summary so I replaced it with an actual summary

> We also set up the Snowman project in our `snowman_project` directory and started writing functions for the Snowman game in our `snowman_project` directory. We will continue adding to our game in future lessons.

Reviewer: ```suggestion
We also set up the Snowman project in our `snowman_project` directory and started writing functions for the Snowman game in `snowman.py`. We will continue adding to our game in future lessons.
```
